server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -519,7 +531,7 @@ public void messageReceived(FieldCapabilitiesNodeRequest request, TransportChann
                 final Map<String, List<ShardId>> groupedShardIds = request.shardIds()
                     .stream()
                     .collect(Collectors.groupingBy(ShardId::getIndexName));
-                final FieldCapabilitiesFetcher fetcher = new FieldCapabilitiesFetcher(indicesService);
+                final FieldCapabilitiesFetcher fetcher = new FieldCapabilitiesFetcher(indicesService, request.includeFieldsWithNoValue());
```

Comment:
I don't recall anymore: we said that the feature requires going to all shards, like index_filter does? Where do we make that decision? What happens if you provide the new flag without an index_filter?

Reply:
My understanding is that all the _field_caps requests are going to all the shards. The current behaviour is to send from the coordinating node one request per index to all the data nodes which in turn sends a request to all the shards that have such index. Just the first response from a shard for a particular index is then accepted in a data node and one node response for each index is accepted in the coordinating node while all the other pending requests are cancelled or discarded. 
The change in behaviour I introduced is to accept all responses in both stages ([here](https://github.com/piergm/elasticsearch/blob/3ab752940aa1f8a0c848bfe017ca782bd14f46a8/server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java#L206-L212) and [here](https://github.com/piergm/elasticsearch/blob/3ab752940aa1f8a0c848bfe017ca782bd14f46a8/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L551-L557)) and merging ([here](https://github.com/piergm/elasticsearch/blob/3ab752940aa1f8a0c848bfe017ca782bd14f46a8/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L170-L180)) the index responses at the data node level in case we have multiple responses for one index.

Reply:
@original-brownbear is what I said correct?

Answering your question:
> What happens if you provide the new flag without an index_filter

I have tests for requests with and without index_filter and the behaviour is consistent in both cases.

Reply:
Maybe your tests don't have enough data nodes or shards to see issues, but I am pretty sure that field caps without index filter will pick one shard per index, instead of all shards for every index. Can you double check?

Reply:
After further analysis all code paths **_can_** go to all shards and all but my code path (return non-empty-fields only) [breaks](https://github.com/piergm/elasticsearch/blob/3ab752940aa1f8a0c848bfe017ca782bd14f46a8/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L556) out of [this loop](https://github.com/piergm/elasticsearch/blob/3ab752940aa1f8a0c848bfe017ca782bd14f46a8/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L539-L564) when they find the first canMatch shard. When no index_filter is passed the first responding shard is always canMatch and breaks directly (unless there is an error), while when an index_filter is provided that could not be the case and the loop iterates to other shards until one returns a canMatch.
That said we should be good, in both cases (with and without index_filter) if my flag is provided we do not break out of the loop and therefore collect all the field capabilities from all the shards.
Another safe measure in place to avoid going to all the shards is an early cached return if the mapping hash is already present in a map, we are covered also there since for non-empty-field we calculate a unique shard-mapping hash.

Reply:
cool thanks for the update. Could you check that we have test coverage about the new flag used in combination with index_filter, as well as not? Thanks!

Reply:
In `FieldCapsHasValueTests` we have tests both with and without index_filter, so we should be good